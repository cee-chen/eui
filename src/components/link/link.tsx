/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, {
  forwardRef,
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
} from 'react';
import classNames from 'classnames';

import { RenderLinkOrButton, useEuiTheme } from '../../services';
import { CommonProps } from '../common';

import { EuiExternalLinkIcon } from './external_link_icon';
import { euiLinkStyles } from './link.styles';

export const COLORS = [
  'primary',
  'subdued',
  'success',
  'accent',
  'danger',
  'warning',
  'text',
  'ghost',
] as const;

export type EuiLinkColor = (typeof COLORS)[number];

export type EuiLinkProps = CommonProps &
  Pick<ButtonHTMLAttributes<HTMLButtonElement>, 'type' | 'disabled'> &
  Omit<AnchorHTMLAttributes<HTMLAnchorElement>, 'type' | 'color'> & {
    /**
     * Any of our named colors.
     */
    color?: EuiLinkColor;
    /**
     * Set to true to show an icon indicating that it is an external link;
     * Defaults to true if `target="_blank"`
     */
    external?: boolean;
  };

export const EuiLink = forwardRef<
  HTMLAnchorElement | HTMLButtonElement,
  EuiLinkProps
>(
  (
    {
      children,
      color = 'primary',
      className,
      href,
      external,
      target,
      rel,
      type = 'button',
      onClick,
      disabled,
      ...rest
    },
    ref
  ) => {
    const euiTheme = useEuiTheme();
    const styles = euiLinkStyles(euiTheme);

    return (
      <RenderLinkOrButton
        className={classNames('euiLink', className)}
        fallbackElement="a"
        elementRef={ref}
        href={href}
        target={target}
        rel={rel}
        isDisabled={disabled}
        onClick={onClick}
        componentCss={(isDisabled) => [
          styles.euiLink,
          isDisabled ? styles.disabled : styles[color],
        ]}
        buttonProps={{ type }}
        linkProps={{
          children: (
            <>
              {children}
              <EuiExternalLinkIcon external={external} target={target} />
            </>
          ),
        }}
        {...rest}
      >
        {children}
      </RenderLinkOrButton>
    );
  }
);

EuiLink.displayName = 'EuiLink';
