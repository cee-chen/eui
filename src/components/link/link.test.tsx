/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { fireEvent } from '@testing-library/react';
import { render } from '../../test/rtl';
import { shouldRenderCustomStyles } from '../../test/internal';
import { requiredProps } from '../../test';

import { EuiLink, COLORS } from './link';

describe('EuiLink', () => {
  shouldRenderCustomStyles(<EuiLink />);

  it('renders', () => {
    const { container } = render(
      <EuiLink href="#" {...requiredProps}>
        <span>Hiya!!!</span>
      </EuiLink>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('color', () => {
    COLORS.forEach((color) => {
      test(`${color} is rendered`, () => {
        const { container } = render(<EuiLink href="#" color={color} />);
        expect(container.firstChild).toMatchSnapshot();
      });
    });
  });

  it('supports both href and onClick', () => {
    const handler = jest.fn();
    const { container } = render(<EuiLink href="/imalink" onClick={handler} />);
    fireEvent.click(container.firstChild!);

    expect(container.firstChild).toHaveAttribute('href', '/imalink');
    expect(handler).toHaveBeenCalledTimes(1);
  });

  it('supports target', () => {
    const { container } = render(<EuiLink href="#" target="_parent" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('supports rel', () => {
    const { container } = render(<EuiLink href="hoi" rel="stylesheet" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('supports disabled', () => {
    const { container } = render(<EuiLink disabled onClick={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('external links', () => {
    it('renders an icon and extra rel', () => {
      const { container } = render(<EuiLink href="#" external />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('is set automatically if `target="_blank"`', () => {
      const { container } = render(<EuiLink href="#" target="_blank" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('allows for target and external to be controlled independently', () => {
      const { container } = render(
        <EuiLink href="#" target="_blank" external={false} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('button behavior', () => {
    it('renders a button if href is not specified and onClick is passed', () => {
      const { container } = render(<EuiLink onClick={() => {}} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('respects the type property', () => {
      const { container } = render(
        <EuiLink type="submit" onClick={() => {}} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('calls onClick', () => {
      const handler = jest.fn();
      const { container } = render(<EuiLink onClick={handler} />);
      fireEvent.click(container.firstChild!);
      expect(handler).toHaveBeenCalledTimes(1);
    });
  });
});
