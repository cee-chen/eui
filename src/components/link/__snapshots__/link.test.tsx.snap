// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiLink button behavior renders a button if href is not specified and onClick is passed 1`] = `
<button
  class="euiLink emotion-euiLink-primary"
  type="button"
/>
`;

exports[`EuiLink color accent is rendered 1`] = `
<a
  class="euiLink emotion-euiLink-accent"
  href="#"
  rel="noreferrer"
/>
`;

exports[`EuiLink color danger is rendered 1`] = `
<a
  class="euiLink emotion-euiLink-danger"
  href="#"
  rel="noreferrer"
/>
`;

exports[`EuiLink color ghost is rendered 1`] = `
<a
  class="euiLink emotion-euiLink-ghost"
  href="#"
  rel="noreferrer"
/>
`;

exports[`EuiLink color primary is rendered 1`] = `
<a
  class="euiLink emotion-euiLink-primary"
  href="#"
  rel="noreferrer"
/>
`;

exports[`EuiLink color subdued is rendered 1`] = `
<a
  class="euiLink emotion-euiLink-subdued"
  href="#"
  rel="noreferrer"
/>
`;

exports[`EuiLink color success is rendered 1`] = `
<a
  class="euiLink emotion-euiLink-success"
  href="#"
  rel="noreferrer"
/>
`;

exports[`EuiLink color text is rendered 1`] = `
<a
  class="euiLink emotion-euiLink-text"
  href="#"
  rel="noreferrer"
/>
`;

exports[`EuiLink color warning is rendered 1`] = `
<a
  class="euiLink emotion-euiLink-warning"
  href="#"
  rel="noreferrer"
/>
`;

exports[`EuiLink external links allows for target and external to be controlled independently 1`] = `
<a
  class="euiLink emotion-euiLink-primary"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <span
    class="emotion-euiScreenReaderOnly"
  >
    (opens in a new tab or window)
  </span>
</a>
`;

exports[`EuiLink external links is set automatically if \`target="_blank"\` 1`] = `
<a
  class="euiLink emotion-euiLink-primary"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <span
    class="emotion-EuiExternalLinkIcon"
    data-euiicon-type="popout"
  >
    External link
  </span>
  <span
    class="emotion-euiScreenReaderOnly"
  >
    (opens in a new tab or window)
  </span>
</a>
`;

exports[`EuiLink external links renders an icon and extra rel 1`] = `
<a
  class="euiLink emotion-euiLink-primary"
  href="#"
  rel="noreferrer"
>
  <span
    class="emotion-EuiExternalLinkIcon"
    data-euiicon-type="popout"
  >
    External link
  </span>
</a>
`;

exports[`EuiLink renders 1`] = `
<a
  aria-label="aria-label"
  class="euiLink testClass1 testClass2 emotion-euiLink-primary-euiTestCss"
  data-test-subj="test subject string"
  href="#"
  rel="noreferrer"
>
  <span>
    Hiya!!!
  </span>
</a>
`;

exports[`EuiLink supports disabled 1`] = `
<button
  class="euiLink emotion-euiLink-disabled"
  disabled=""
  type="button"
/>
`;
